### PR TITLE
Bugfix/custom dataset embeddeding pfs

### PR DESCRIPTION
**Pull request by anwar6953**
https://issues.cask.co/browse/CDAP-3030
http://builds.cask.co/browse/CDAP-RBT349-6

The issue described in the JIRA is that the dependencies declared by a module using the module partitionedFileSet (pfs) or the module timePartitionedfileSet (tpfs) did not (in pre-3.1 CDAP) include 'core', which the pfs and tpfs now depend on (3.1 and following).

This stage in the upgrade tool modifies any modules that depend on pfs or tpfs to also declare a dependency on 'core'.

### CODE REVIEW

**Comment by anew** [comment]
LGTM if tested on cluster

**Comment by anwar6953** [comment]
Added another commit because the order of `usesModules` matters in terms of the order in which the dependencies are loaded.

File: cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java [hunk 1]
```diff
@@ -149,8 +153,63 @@ public void apply() throws Exception {
...
+      }
+    }
+    return false;
+  }
+
+  @VisibleForTesting
+  DatasetModuleMeta migrateDatasetModuleMeta(DatasetModuleMeta moduleMeta) {
+    List<String> newUsesModules = new ArrayList<>(moduleMeta.getUsesModules());
+    // core needs to be to the left of pfs or tpfs, because this list is ordered by dependency
+    newUsesModules.add(0, "core");
```
Comment (albertshau) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java]: more efficient in array list to add this first, then add the rest.

File: cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java [hunk 2]
```diff
@@ -149,8 +153,63 @@ public void apply() throws Exception {
...
+          DatasetModuleMeta migratedModuleMeta = migrateDatasetModuleMeta(moduleMeta);
+          dsTypeMDS.write(entry.getKey(), migratedModuleMeta);
+        }
+      }
+    });
+  }
+
+  @VisibleForTesting
+  boolean needsConverting(DatasetModuleMeta moduleMeta) {
+    for (String usesModule : moduleMeta.getUsesModules()) {
```
Comment (albertshau) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java]: does this catch datasets that depend on a module that depends on pfs or tpfs?

Comment (anwar6953) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java]: yea, they're all listed here. I don't think the dependency declaration is recursive (all are declared explicitly).

File: cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java [hunk 3]
```diff
@@ -149,8 +153,64 @@ public void apply() throws Exception {
...
+  }
 
+  @VisibleForTesting
+  boolean needsConverting(DatasetModuleMeta moduleMeta) {
+    // if 'core' is already a declared dependency, no need to upgrade the module meta
+    List<String> usesModules = moduleMeta.getUsesModules();
+    if (usesModules.contains("core")) {
+      return false;
+    }
+    return usesModules.contains("partitionedFileSet") || usesModules.contains("timePartitionedFileSet");
```
Comment (albertshau) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java]: technically somebody could have used "partitionedFileSet" or ```PartitionedFileSet.class.getName()```

Comment (albertshau) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java]: same thing for tpfs

File: cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java (hunk 2 above)
Comment (anwar6953) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java]: this is the reason we need to add 'core' to modules that depend on pfs/tpfs, even though the pfs/tpfs modules depend on 'core' in their module meta.

File: cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java (hunk 3 above)
Comment (anew) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java]: It is looking for a module name, not a dataset type name. The module name is unique

Comment (anwar6953) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/PFSUpgrader.java]: yea, I looked at the dataset.type table, and concluded the same thing as Andreas.

**Comment by anwar6953** [comment]
Also added a dependency on 'orderedTable-hbase'.

**Comment by anwar6953** [comment]
Good point.

**Comment by anwar6953** [comment]
Pushed latest commit with updates.
@anew @anew
Please take another look; thanks.

**Comment by anew** [comment]
LGTM. Please test before merging. This turned out to be trickier than expected...

**Comment by anwar6953** [comment]
@anew @albertshau 
Pushed another commit which resolves an issue which happens when trying to migrate the tables of an embedded pfs. Explanation is in the comment of the commit.

**Comment by anew** [comment]
LGTM after testing :)

**Comment by anwar6953** [comment]
Tested both top-level pfs as well as an embedded tpfs, retrieval of partitions created before the upgrade was succesfull in both cases now.

**Comment by anew** [comment]
merge it!

**Pull request merged**